### PR TITLE
Fix "Play Offline" issues

### LINKED ISSUE #85: Play Offline not working with MSA when servers are offline
### Operating System

Windows, macOS, Linux, Other

### Description of bug

When `https://api.minecraftservices.com/launcher/login` isn't reachable, *Play offline* doesn't work when logged in with MSA.

This dialog closes and re-appears every now and then.

```
   15.379 D "Mojang" is valid.
   15.379 D AuthFlow: "Accessing Mojang services."
   15.379 D Getting Minecraft access token...
   25.589 W AuthRequest::onRequestError: Error 499
   25.589 W AuthRequest::onRequestError: Error string:  "Error transferring https://api.minecraftservices.com/launcher/login - server replied: Gateway Time-out"
   25.590 W AuthRequest::onRequestError: HTTP status 504 "Gateway Time-out"
   25.590 D "<html>\r\n<head><title>504 Gateway Time-out</title></head>\r\n<body>\r\n<center><h1>504 Gateway Time-out</h1></center>\r\n</body>\r\n</html>\r\n"
   25.590 W Reply error: QNetworkReply::UnknownServerError
   25.590 D "<html>\r\n<head><title>504 Gateway Time-out</title></head>\r\n<body>\r\n<center><h1>504 Gateway Time-out</h1></center>\r\n</body>\r\n</html>\r\n"
   25.590 C Task "MSASilent(0x562df882aed0)" failed:  "Failed to get Minecraft access token: Error transferring https://api.minecraftservices.com/launcher/login - server replied: Gateway Time-out"
```

Related https://github.com/MultiMC/Launcher/issues/4433

### Steps to reproduce

1. Make sure `https://api.minecraftservices.com/launcher/login` isn't reachable
2. Open PolyMC
3. Click *Play offline* for any instance

### Suspected cause

_No response_

### This issue is unique

- [X] I have searched the issue tracker and did not find an issue describing my bug.

### PR DESCRIPTION
**Pull request by Scrumplex**
This PR removes offline accounts, and it fixes #85 and possibly other related errors.

### CODE REVIEW

File: launcher/net/NetUtils.h [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include <qnetworkreply.h>
```
Comment (flowln) [inline launcher/net/NetUtils.h]:
```suggestion
```

File: launcher/net/NetUtils.h [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+#include <QNetworkReply>
+
+namespace Net {
+    inline bool isApplicationError(QNetworkReply::NetworkError x) {
+        // Mainly taken from https://github.com/qt/qtbase/blob/dev/src/network/access/qhttpthreaddelegate.cpp
+        QList<QNetworkReply::NetworkError> errors = {
```
Comment (flowln) [inline launcher/net/NetUtils.h]:
```suggestion
        static QSet<QNetworkReply::NetworkError> errors = {
```

i think that's a more suitable data structure for this :)

**Review by flowln — APPROVED** [review]
:+1:

**Review by txtsd — APPROVED** [review]

**Review by ZekeZDev — APPROVED** [review]

File: launcher/net/NetUtils.h [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+        static QSet<QNetworkReply::NetworkError> errors = {
+            QNetworkReply::ProtocolInvalidOperationError,
+            QNetworkReply::AuthenticationRequiredError,
+            QNetworkReply::ContentAccessDenied,
+            QNetworkReply::ContentNotFoundError,
+            QNetworkReply::ContentOperationNotPermittedError,
+            QNetworkReply::ProxyAuthenticationRequiredError,
+            QNetworkReply::ContentConflictError,
+            QNetworkReply::ContentGoneError,
+            QNetworkReply::InternalServerError,
```
Comment (LennyMcLennington) [inline launcher/net/NetUtils.h:36]: What's the difference between STATE_FAILED_SOFT and STATE_OFFLINE, and why wouldn't it emit STATE_OFFLINE when there's an internal server error?

**Review by LennyMcLennington — CHANGES_REQUESTED** [review]

**Review by LennyMcLennington — CHANGES_REQUESTED** [review]
This removes offline accounts but I don't see how it fixes all the cases where accounts are marked as 'expired' despite not being expired. Sure, it does fix other issues and that's good but I don't think it's okay to remove offline accounts until that's fixed.

Please don't remove offline accounts until it's no longer possible for accounts to be marked as expired when they're perfectly valid.

File: launcher/net/NetUtils.h (hunk 3 above)
Comment (LennyMcLennington) [reply launcher/net/NetUtils.h:36]: nvm

**Comment by Scrumplex** [comment]
@LennyMcLennington The only cases where an account will be set as "Expired" is when Mojang or MS APIs don't return the necessary information for an account, or return an error. If you have experienced that in the past, then some log output would be helpful. Right now I don't see many potential places where that could happen accidentally (see `launcher/minecraft/auth/Yggdrasil.cpp` and `launcher/minecraft/auth/steps/MSAStep.cpp`, look for `AccountTaskState::STATE_FAILED_HARD`).

In any case, "Play Offline" always worked like this. Offline accounts just don't check this kind of stuff at all.

**Comment by Scrumplex** [comment]
Moving the removal of offline accounts to a different milestone

**Review by LennyMcLennington — APPROVED** [review]

**Comment by DioEgizio** [comment]
4 approvals 👀

**Pull request merged**